docs/development/APPLICATION_DATA.md

```diff
@@ -0,0 +1,148 @@
+
+When sending or receiving application data, we must read or write that data as encrypted TLS records.
+
+## Fragmentation
+
+TLS records may only contain a limited amount of data. If an application wants to send a large amount of data, that data will be broken into "fragments" and each fragment will be sent in a separate record.
+
+The maximum fragment length is defined by the TLS protocol as 2^14. s2n-tls uses a default fragment length of about 8k, which a code comment explains as "Testing in the wild has found 8k max record sizes give a good balance of low latency and throughput." Most TLS libraries default to about 16k.
+
+The client and server can use the "Maximum fragment length" extension to negotiate a different fragment length. Lowering the fragment length can improve latency, since each individual record will be faster to process (see s2n_connection_prefer_low_latency). Raising the fragment length can improve throughput, since fewer records will need to be processed (see s2n_connection_prefer_throughput).
+
+Changing the fragment length changes how s2n-tls constructs records, but does NOT change how it parses records. s2n-tls will accept record payloads that exceed the negotiated maximum fragment length. Historically TLS implementations have not always respected the negotiated maximum fragment length, so rejecting oversized records could lead to compatibility issues with older TLS libraries.
+
+### Dynamic Record Sizes
+
+s2n-tls also supports a feature where the fragment size starts small and grows to the maximum over a specified period of time. See "s2n_connection_set_dynamic_record_threshold".
+
+## Sending Application Data
+
+To send application data, the customer calls "s2n_send". See the customer-facing usage guide. But when the customer calls s2n_send, what actually happens?
+
+### s2n_sendv and s2n_sendv_with_offset
+
+This document will only refer to "s2n_send", but also covers the other send methods. The methods only differ in interface and user experience. "s2n_send" and "s2n_sendv" are just wrappers around "s2n_sendv_with_offset".
+
+### The out buffer
+
+A record is staged in the `out` buffer before being written to the network. Both the headers and the payloads are written to `out`, so it contains complete records. The records are encrypted in-place.
+
+By default, `out` is allocated to the maximum record size given the maximum fragment size, so can only contain a single record at a time. This can be configured: see [Multi-record send](#multi-record-send) below.
+
+### The initial send attempt
+
+The first time the application attempts to send a given chunk of data, the basic process is fairly simple. s2n-tls will break the data into fragments, copy each fragment into the `out` buffer with a record header, encrypt the record, and send the record over the network.
+
+However, if sending any of the records over the network fails, then the call to s2n_send will need to be retried.
```

Comment:
```suggestion
However, if sending any of the records over the network fails, then the user will need to call s2n_send again.
```
Just want to make it clear that we(s2n-tls) do not retry the send call. We just report back what we successfully sent and it is the users responsibility to call s2n_send again.